packages/devkit/src/generators/plugin-migrations/plugin-migration-utils.ts

```diff
@@ -0,0 +1,521 @@
+    );
+  }
+
+  let project: ProjectConfiguration;
+  if (options.project) {
+    project = readProjectConfiguration(tree, options.project);
+  }
+
+  const projectFilter = project?.name
+    ? (tree: Tree, allProjectsWithExecutor: Map<string, Set<string>>) => {
+        if (project?.name && !allProjectsWithExecutor.has(project?.name)) {
+          throw new Error(
+            `Project "${project?.name}" does not use "${executor}" executor. Please select a project that does."`
+          );
+        }
+
+        let filteredProjects = new Set(allProjectsWithExecutor.keys());
+        if (pluginProjectFilter) {
+          filteredProjects = pluginProjectFilter(tree, allProjectsWithExecutor);
+        }
+
+        if (!filteredProjects.has(project.name)) {
+          throw new Error(`Project "${project.name} cannot be migrated."`);
+        }
+
+        return new Set<string>([project.name]);
+      }
+    : pluginProjectFilter;
+
+  let projectsLeftToMigrate: Set<string> = new Set();
+  do {
+    const { projects, targetName, projectsStillToBeMigrated } =
+      getProjectsToMigrate(tree, executor, projectFilter);
+
+    if (projects) {
+      const { include, migratedProjects } = await migrateProjectsToPlugin<T>(
+        tree,
+        projectGraph,
+        projects,
+        targetName,
+        createProjectsConfig,
+        executor,
+        createNodes,
+        postTargetTransformer
+      );
+
+      addPluginWithOptions<T>(
+        tree,
+        pluginPath,
+        include,
+        createPluginOptions(targetName)
+      );
+    }
+
+    projectsLeftToMigrate = projectsStillToBeMigrated;
+  } while (projectsLeftToMigrate.size !== 0);
+}
+
+export async function migrateProjectsToPlugin<T>(
+  tree: Tree,
+  projectGraph: ProjectGraph,
+  projects: string[],
+  targetName: string,
+  createProjectsConfig: (
+    tree: Tree,
+    projectRoot: string,
+    targetName: string,
+    context: CreateNodesContext
+  ) => CreateNodesResult | Promise<CreateNodesResult>,
+  executor: string,
+  createNodes: CreateNodes<T>,
+  postTargetTransformer: (
+    target: TargetConfiguration,
+    tree?: Tree,
+    projectRoot?: string
+  ) => TargetConfiguration = (targetConfiguration) => targetConfiguration
+): Promise<{
+  targetName: string;
+  include: string[] | undefined;
+  migratedProjects: string[];
+}> {
+  const nxJsonConfiguration = readNxJson(tree);
+  const targetDefaultsForExecutor =
+    nxJsonConfiguration.targetDefaults?.[executor];
+
+  const configFiles = glob(tree, [createNodes[0]]);
+  let include: string[] = [];
+  const migratedProjects = [];
+  for (const projectName of projects) {
+    const projectFromGraph = projectGraph.nodes[projectName];
+
+    const context = {
+      workspaceRoot: tree.root,
+      nxJsonConfiguration,
+      configFiles,
+    };
+    const projectConfigurations = await createProjectsConfig(
+      tree,
+      projectFromGraph.data.root,
+      targetName,
+      context
+    );
```

Comment:
This is weird... keep it for now... we should remove this ideally before we do too many plugins.

Let's get an initial version in.. and start using it.